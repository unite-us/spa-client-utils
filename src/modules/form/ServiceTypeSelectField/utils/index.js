import filterServiceTypeOptions from './filterServiceTypeOptions';

export {
  filterServiceTypeOptions,
};
