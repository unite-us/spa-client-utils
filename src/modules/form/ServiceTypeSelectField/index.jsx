import React, { Component } from 'react';
import _ from 'lodash';
import PropTypes from 'prop-types';
import { SelectField } from '@unite-us/ui';
import { filterServiceTypeOptions } from './utils';

class ServiceTypeSelectField extends Component {
  constructor(props) {
    super(props);

    this.loadOptions = this.loadOptions.bind(this);
    this.resetOptions = this.resetOptions.bind(this);
  }

  loadOptions(search) {
    const { options } = this.props;
    const filtered = filterServiceTypeOptions(options, search);
    return new Promise((resolve) => {
      resolve({
        options: filtered,
      });
    });
  }

  resetOptions() {
    const { field, touch } = this.props;
    // Mark the field as touched.
    touch(field.name);
    // Clear search text value.
    this.selectField.choices.input.value = '';
    // Reset options.
    this.selectField.onSearch({ detail: { value: '' } });
  }

  render() {
    const { field, registerField, ...props } = this.props;

    return (
      <SelectField
        {...(_.isEmpty(field) ? {} : {
          field: {
            ...field,
            onBlur: this.resetOptions,
          },
        })}
        loadOptions={this.loadOptions}
        ref={(n) => {
          this.selectField = n;
          registerField(n);
        }}
        {...props}
      />
    );
  }
}

ServiceTypeSelectField.propTypes = {
  className: PropTypes.string,
  disabled: PropTypes.bool.isRequired,
  field: PropTypes.object.isRequired,
  id: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  options: PropTypes.array,
  placeholder: PropTypes.string,
  registerField: PropTypes.func.isRequired,
  searchChoices: PropTypes.bool,
  searchPlaceholderValue: PropTypes.string,
  touch: PropTypes.func.isRequired,
  validations: PropTypes.oneOfType([
    PropTypes.object,
    PropTypes.array,
  ]).isRequired,
};

ServiceTypeSelectField.defaultProps = {
  className: '',
  id: '',
  inline: false,
  label: 'Service Type',
  labelKey: 'name',
  onChange: _.noop,
  options: [],
  placeholder: '',
  searchChoices: true,
  searchPlaceholderValue: '',
  valueKey: 'id',
};

export default ServiceTypeSelectField;
