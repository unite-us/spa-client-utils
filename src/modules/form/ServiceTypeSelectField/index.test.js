import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import ServiceTypeSelectField from './index';

const props = {
  disabled: false,
  field: {},
  placeholder: 'placeholder text',
  options: [
    {
      name: 'Category 1',
      children: [
        { name: 'benefits for category 1' },
        { name: 'education for category 2' },
      ],
    },
    {
      name: 'Category 3',
      children: [
        { name: 'Moving assistance for category 3' },
        { name: 'Taxes for category 3' },
      ],
    },
  ],
  searchChoices: false,
  className: '',
  id: '',
  registerField: jest.fn(),
  searchPlaceholderValue: 'placeholder value',
  touch: jest.fn(),
  validations: {},
};


describe('ServiceTypeSelectField', () => {
  it('renders', () => {
    const createNodeMock = (element) => {
      if (element.type === 'select') {
        return {
          addEventListener: jest.fn(),
        };
      }
      return null;
    };

    const options = { createNodeMock };
    expect(renderer.create(<ServiceTypeSelectField {...props} />, options)).toMatchSnapshot();
  });

  it('loads correct filter options when search is entered', () => {
    const wrapper = shallow(<ServiceTypeSelectField {...props} />);
    const search = 'Category 3';
    const expected = { options: [props.options[1]] };

    return wrapper.instance().loadOptions(search).then((data) => {
      expect(data).toEqual(expected);
    });
  });

  it('resets options', () => {
    const createNodeMock = (element) => {
      if (element.type === 'select') {
        return {
          addEventListener: jest.fn(),
        };
      }
      return null;
    };

    const options = { createNodeMock };
    const instance = renderer.create(<ServiceTypeSelectField {...props} />, options);
    const selectField = instance.getInstance().selectField;
    const onSearch = jest.spyOn(selectField, 'onSearch');

    selectField.choices.input = { value: 'foo' };

    instance.getInstance().resetOptions();

    expect(selectField.choices.input.value).toEqual('');
    expect(onSearch).toHaveBeenCalledWith({ detail: { value: '' } });
    expect(props.touch).toHaveBeenCalled();
  });
});
