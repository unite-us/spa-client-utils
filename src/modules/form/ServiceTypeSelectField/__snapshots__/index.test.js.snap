// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServiceTypeSelectField renders 1`] = `
<div
  className="ui-select-field ui-form-field"
  style={Object {}}
>
  <label
    className="ui-form-field__label"
    htmlFor=""
    style={Object {}}
  >
    Service Type
  </label>
  <div
    style={
      Object {
        "position": "relative",
      }
    }
  >
    <div
      className=""
      data-test-element="dropdown_undefined"
    >
      <select
        id=""
        multiple={false}
      />
    </div>
    <div
      className="ui-form-field__hint"
      style={Object {}}
    >
      
    </div>
    <div
      className="ui-form-field__error"
      style={Object {}}
    />
  </div>
</div>
`;
